scl/cri/cri.conf

```diff
@@ -0,0 +1,51 @@
+#########################################################
+# Parser for CRI log format
+# ISO_TIME(nano) SP STREAM SP TAG SP MSG
+# https://github.com/kubernetes/kubernetes/issues/44976
+#========================================================
+# Retrieves information about K8S meta from log filename:
+#  - container ID
+#  - container Name
+#  - POD NameSpace
+#  - POD Name
+# Use .cri prefix
+# Set .cri.time as S_ISODATE with use of date-parser()
+#
+# Note: does not support multiline, line = message
+#########################################################
+# Example log:
+# 2022-02-04T18:14:43.219493781+01:00 stdout F Starting up on port 80
+
+@define crio-log-regexp "([a-z0-9-]+)_([a-z0-9-]+)_([a-z0-9]+)-([a-z0-9]{64}).log"
+
+block rewrite extract-container-id (value("MESSAGE")){
+    # make the container-id of 12 characters long as usual in cli
+    subst("`crio-log-regexp`", "$(substr $4 0 12)", value("`value`"), flags(global, store-matches), type(pcre));
+};
+
+block parser cri-parser(prefix('.cri.')) {
+    channel {
+        parser {
+            csv-parser(
+                columns("`prefix`time", "`prefix`stream", "`prefix`flag", "`prefix`message")
+                delimiters(" ")
+                flags(greedy)
+            );
+            # set S_ISODATE to value of ${.cri.time}
+            date-parser(format("%FT%H:%M:%S.%f%Z") template("${.cri.time}"));
+        };
+	# drop everything which does not match (to exclude logs)
+        #filter { not match("" value("`prefix`COMMAND") type(string)); };
+        rewrite {
+            set("$(basename \"${FILE_NAME}\")", value("`prefix`container.id"));
```

Comment:
This is only applied to a file source?

Maybe in that case the "syslog" topic is not the right one.

Apart from a sample log line can you post a complete config?

Or at least explain how these messages get to syslog-ng? Thanks

Reply:
Yes. It's file source.
Example source config:
https://github.com/peterviskup/syslog-ng-container/blob/main/syslog-ng/conf.d/10-source-containers.conf
parser call
https://github.com/peterviskup/syslog-ng-container/blob/main/syslog-ng/conf.d/20-store-containers.conf
and template for syslog forward
https://github.com/peterviskup/syslog-ng-container/blob/main/syslog-ng/conf.d/09-templates.conf#L5
and Kubectl DeamonSet for syslog-ng with /var/log/containers and /var/log/pods paths available
https://github.com/peterviskup/syslog-ng-container/blob/main/kubernetes/syslog-ng-daemonset.yaml

Reply:
Oh, I didn't  know your syslog-ng-container initiative but it is really exciting.

Normally, each "topic" in app-parser identifies a means how log is entered syslog-ng, and reading application specific logfiles fits the bill of "means-of-getting-logs". And I want to make this as out-of-the-box as possible. The topic makes it possible to add application adapters, e.g. configuration snippets that classify and parse application specific originating from a specific topic.

The default-network-drivers() source encapsulates the best practice how someone should consume syslog in a number of transports. This driver defines "syslog" and "syslog-raw" topics. "syslog-raw" as a topic can be used for non-standard logs, Cisco for instance breaks syslog in a number of interesting ways. So the application adapter for cisco logs taps into syslog-raw, where it gets a chance to fix it prior to parsing. "syslog" on the other hand is for log data that is successfully  parsed.

With that in mind, I I would use a separate topic for logs coming from /var/log/container, especially if they are in a specific format (which seemingly is).

And the source to encapsulate your file-source could be a source block of its own:

```
block source local-containers() {
  wildcard-file(
    base-dir("/var/log/containers")
    filename-pattern("*.log")
    recursive(no)
    follow-freq(1)
    flags(no-parse)
  );
  channel {
      parser { app-parser(topic(containers-log)); };
  };
};
```

Then the user is able to use your snippet this way:

```
source s_whatever {
    system();
    default-network-drivers();
    local-containers();
};
```

This way the user has freedom to define her sources, while the entire logic of fetching the logs/classifying/parsing them is
encapsulated in the local-containers() block and the application adapters "subscribing" to the "containers-log" topic.

Of course the naming is probably not right and could be improved above, also some of the parameters might need to
be extracted so that the user can override it, but I hope you get the idea.

This might not match exactly what you are trying to achieve. Can we perhaps jump to our gitter.im channel for a chat? I would like to understand your use-case and see if the current app-parser() model applies entirely or it needs to be adapted. Thanks.

Reply:
this seems way better and does make sense
of course, we can have a chat